File: smalivm/src/main/java/org/cf/smalivm/opcode/BinaryMathOpFactory.java
package org.cf.smalivm.opcode;

import gnu.trove.map.TIntObjectMap;

import org.cf.smalivm.ExceptionFactory;
import org.cf.smalivm.VirtualMachine;
import org.cf.util.Utils;
import org.jf.dexlib2.builder.BuilderInstruction;
import org.jf.dexlib2.builder.MethodLocation;
import org.jf.dexlib2.iface.instruction.NarrowLiteralInstruction;
import org.jf.dexlib2.iface.instruction.TwoRegisterInstruction;
import org.jf.dexlib2.iface.instruction.formats.Instruction12x;
import org.jf.dexlib2.iface.instruction.formats.Instruction23x;

public class BinaryMathOpFactory implements OpFactory {

    @Override
    public BinaryMathOp create(MethodLocation location, TIntObjectMap<MethodLocation> addressToLocation,
                               VirtualMachine vm) {
        MethodLocation child = Utils.getNextLocation(location, addressToLocation);
        BuilderInstruction instruction = (BuilderInstruction) location.getInstruction();
        TwoRegisterInstruction instr = (TwoRegisterInstruction) location.getInstruction();
        int destRegister = instr.getRegisterA();
        int arg1Register = instr.getRegisterB();
        ExceptionFactory exceptionFactory = vm.getExceptionFactory();
        if (instruction instanceof Instruction23x) {
            // add-int vAA, vBB, vCC
            int arg2Register = ((Instruction23x) instruction).getRegisterC();
            return new BinaryMathOp(location, child, destRegister, arg1Register, arg2Register, false, exceptionFactory);
        } else if (instruction instanceof Instruction12x) {
            // add-int/2addr vAA, vBB
            arg1Register = instr.getRegisterA();
            int arg2Register = ((Instruction12x) instruction).getRegisterB();
            return new BinaryMathOp(location, child, destRegister, arg1Register, arg2Register, false, exceptionFactory);
        } else if (instruction instanceof NarrowLiteralInstruction) {
            // Instruction22b - add-int/lit8 vAA, vBB, #CC
            // Instruction22s - add-int/lit16 vAA, vBB, #CCCC
            int arg2Literal = ((NarrowLiteralInstruction) instruction).getNarrowLiteral();
            return new BinaryMathOp(location, child, destRegister, arg1Register, arg2Literal, true, exceptionFactory);
        } else {
            return null;
        }
    }

}

File: smalivm/src/main/java/org/cf/smalivm/opcode/BinaryMathOpFactory.kt
package org.cf.smalivm.opcode

import gnu.trove.map.TIntObjectMap
import org.cf.smalivm.VirtualMachine
import org.cf.util.Utils
import org.jf.dexlib2.builder.BuilderInstruction
import org.jf.dexlib2.builder.MethodLocation
import org.jf.dexlib2.iface.instruction.NarrowLiteralInstruction
import org.jf.dexlib2.iface.instruction.TwoRegisterInstruction
import org.jf.dexlib2.iface.instruction.formats.Instruction12x
import org.jf.dexlib2.iface.instruction.formats.Instruction23x

class BinaryMathOpFactory : OpFactory {
    fun create(
        location: MethodLocation, addressToLocation: TIntObjectMap<MethodLocation?>?,
        vm: VirtualMachine
    ): BinaryMathOp? {
        val child = Utils.getNextLocation(location, addressToLocation)
        val instruction = location.instruction as BuilderInstruction?
        val instr = location.instruction as TwoRegisterInstruction?
        val destRegister = instr!!.registerA
        var arg1Register = instr.registerB
        val exceptionFactory = vm.exceptionFactory
        return if (instruction is Instruction23x) {
            // add-int vAA, vBB, vCC
            val arg2Register = (instruction as Instruction23x).registerC
            BinaryMathOp(location, child, destRegister, arg1Register, arg2Register, false, exceptionFactory)
        } else if (instruction is Instruction12x) {
            // add-int/2addr vAA, vBB
            arg1Register = instr.registerA
            val arg2Register = (instruction as Instruction12x).registerB
            BinaryMathOp(location, child, destRegister, arg1Register, arg2Register, false, exceptionFactory)
        } else if (instruction is NarrowLiteralInstruction) {
            // Instruction22b - add-int/lit8 vAA, vBB, #CC
            // Instruction22s - add-int/lit16 vAA, vBB, #CCCC
            val arg2Literal = (instruction as NarrowLiteralInstruction).narrowLiteral
            BinaryMathOp(location, child, destRegister, arg1Register, arg2Literal, true, exceptionFactory)
        } else {
            null
        }
    }
}
